# Ponder and Prove Combinatorics and Probability
#### Due: Saturday, 6 February 2021, 11:59 pm.

## Conjecture

A number-theoretic conjecture of combinatorial significance is the following:

$degree2({2n \choose n}) =$ the "bits-on count" (or population count, or Hamming weight) of $n$.

$degree2(m)$ is defined as the number (degree, exponent) of 2's in the prime factorization of $m$.

In other words, for any $m$, a positive integer, $m = 2^e \cdot o$ where $o$ is an odd positive integer (could be 1) and $e$ is a natural number, including zero --- which would be the case when $m$ is odd. It's the $e$ that is the $degree2$ of $m$.

Another way to state this conjecture is that the number of 1's in the binary expansion of ${2n \choose n}$ for positive integer $n$ is equal to the number of 2's in the prime factorization of $n$.

Your task is to write Python code to test this conjecture for as many positive integers as you can. See the self-assessment for more details.

Note: a `bitsoncount` function can be a one-liner in Python: `return bin(x).count('1')`



In [7]:
from math import gcd

def nCk(n, k):
  if k < 0 or k > n:
    return 0
  else:
    result = 1
    d = 1
    g = 1
    m = min(k, n - k)
    while (d <= m):
      g = gcd(result, d)
      result = n * (result // g)
      result = (result // (d // g))
      n -= 1
      d += 1
    return result

def bitsoncount(x):
    return bin(x).count('1')

# https://www.geeksforgeeks.org/count-occurrences-of-a-prime-number-in-the-prime-factorization-of-every-element-from-the-given-range/
def twosoncount(x):
    prime = 2 # The prime we are looking for
    count = 0 
    val = prime
    while (True): 
        # Number of values in n that are divisible by val 
        a = x // val

        # Number of values in n - 1 that are divisible by val 
        b = (x - 1) // val

        # Increment the power of the val 
        val *= prime

        # (a - b) is the count of numbers in n that are divisible by val 
        if (a - b): 
            count += (a - b)

        # No values that are divisible by val 
        # thus exiting from the loop 
        else: 
            break

    return int(count)


def test_conjecture(n):
    return bitsoncount(n) == twosoncount(nCk(2*n, n))

In [8]:
import datetime

print('Beginning conjecture verification at ' + str(datetime.datetime.now()))

n = 1
try:
  while test_conjecture(n):
    n += 1
except:
  print('Verified up to ' + str(n) + ' at ' + str(datetime.datetime.now()))

Beginning conjecture verification at 2021-02-05 23:37:32.376174
Verified up to 3774 at 2021-02-05 23:38:08.989387


## Results

After running the above code on my laptop for 24 hours, I got the following output: 

> Beginning conjecture verification at 2021-02-05 23:37:32.376174

> Verified up to 3774 at 2021-02-05 23:38:08.989387



## Basic Probability Theory Question
A dark room contains two barrels. The first barrel is filled with green marbles, the second is filled with a half-and-half mixture of green and blue marbles. So there's a 100% chance of choosing a green marble from the first barrel, and a 50% chance of choosing either color in the second barrel. You reach into one of the barrels (it's dark so you don't know which one) and select a marble at random. It's green. You select another. It's green too. You select a third, a fourth, a fifth, etc. Green each time. What is the *minimum* number of marbles you need to select to *exceed* a probability of 99% that you are picking them out of the all-green barrel? (Note that there are enough marbles so that the answer does not depend on how many marbles are in the second barrel.)


In [13]:
num_marbles = 1
prob = 0.5
while (prob >= .01):
    num_marbles += 1
    prob *= .5
print(num_marbles)

7


The above code shows that 7 is the minimum number of marbles needed to exceed the specified probability of 99%. 

## A Related But Deeper Basic Probability Theory Question
Take a deep breath. Suppose Shakespeare's account is accurate and Julius Caesar gasped "You too, Brutus" before breathing his last. What is the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath?

Assume that after more than two thousand years the exhaled molecules are uniformly spread about the world and the vast majority are still free in the atmosphere. Assume further that there are $10^{44}$ molecules of air in the world, and that your inhaled quantity and Caesar's exhaled quantity were each about $2.2 \times 10^{22}$ molecules.
### Hint
If a number $x$ is small, then $(1 - x)$ is approximately equal to $e^{-x}$.


In [14]:
from math import e 

molecules = 2.2 * (10**22)
x = molecules / (10**44)
print(1 - e ** (-x * molecules))

0.9920929459484066


The above code shows that the probability of inhaling a molecule that Caesar exhaled is >99%, which is both extraordinary and logical! 

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


### What is true of my experience in general?
(5 points each, 15 points total)
- [x] I had fun.
- [x] I learned something new.
- [x] I achieved something meaningful, or something I can build upon at a later time.

### What is true of my report on what I learned?
(5 points each, 25 points total)
- [x] I wrote a sufficient number of well-written sentences.
- [x] My report is free of "mechanical infelicities" (misspelled words, grammatical errors, punctuation errors, etc.).
- [x] I reported on any connections I found between this investigation and something I already know.
- [x] I reported who were and what contribution each of my collaborators made.
- [x] I reported on how many numbers I was able to verify with a time/computation budget of 24 hours (in a row).


### What is true about my answers?
(15 points each, 60 points total)
- [ ] I figured out how to run a Python program continuously for at least 24 hours.
- [ ] I refrained from printing out anything except the highest number I verified, knowing that printing just slows a program down.
- [ ] I got the right answer for the first probability theory question.
- [ ] I got the right answer for the second probability theory question.
